Geometry_Engine/Modify/SplitAtPoints.cs

```diff
@@ -212,111 +212,113 @@ public static List<PolyCurve> SplitAtPoints(this PolyCurve curve, List<Point> po
+                        if (ptsOnSegment.Any(x => x.SquareDistance(ePt) < sqTol)) //If split at endpoint of current segment
                         {
-                            j++;
-                            break;
+                            result.Add(new PolyCurve { Curves = new List<ICurve> { split.Last() } });   //Add last segment to return list
+                            prev = null;    //Set prev to null
                         }
-                        else if (tmpResult[j].IEndPoint().IsEqual(curve.EndPoint()) || (isClosedAndAntiClockwise && tmpResult[j].IEndPoint().IsEqual(curve.StartPoint())))
+                        else    //If not split at end point of segment
                         {
-                            j++;
-                            break;
+                            prev = new PolyCurve { Curves = new List<ICurve> { split.Last() } };    //Set prev to contain the last split segment
                         }
                     }
-                    j++;
+                }
+                else    //If no splitpoints available
+                {
+                    if (prev != null)   //If prev is set
+                        prev.Curves.Add(crv);   //Add full segment to prev
+                    else
+                        prev = new PolyCurve { Curves = new List<ICurve> { crv } }; //If prev is not set, set it with full segment added
                 }
 
-                if (subResultList.Count > 0)
-                    result.Add(new PolyCurve { Curves = subResultList.ToList() });
-
-                i++;
-            }
-
-            if (curve.IsClosed(tolerance) && !(curve.SubParts()[0] is Circle))
-                if (!curve.StartPoint().IsEqual(onCurvePoints[onCurvePoints.Count - 1]))
+                if (k == subParts.Count - 1)    //Last segment
                 {
-                    List<ICurve> subResultList = new List<ICurve>();
-                    foreach (ICurve subCrv in result[result.Count - 1].ISubParts())
-                        subResultList.Add(subCrv);
-                    foreach (ICurve subCrv in result[0].ISubParts())
-                        subResultList.Add(subCrv);
-                    result.RemoveAt(0);
-                    result.RemoveAt(result.Count - 1);
-                    result.Add(new PolyCurve { Curves = subResultList.ToList() });
+                    if (prev != null)   //If prev is set
+                    {
+                        if (curve.IsClosed(tolerance) && result.Count != 0) //If curve is closed
+                            result[0].Curves.InsertRange(0, prev.Curves);   //Insert the curves in prev to the start of the first curve, ensuring a segment looping around the start/end is added
+                        else
+                            result.Add(prev);   //If not, simply add the prev to the return list
+                    }
                 }
+            }
 
             return result;
```

Comment:
I would unwrap single-segment `PolyCurves` before returning (alternatively you could store all types in `result` while iterating through the segments, but this will require replacing single segments with `PolyCurves` when adding another segment to them)

Reply:
Would require to change the return type of this method from PolyCurve to ICurve though. Not sure if we want to do that?

Reply:
Fair point - I would have nothing against, taken that `PolyCurve` implement `ICurve`. But curious about your opinion on changing the return type?

Reply:
Would not really mind changing it in general, mostly cautios about changing it in case it cascades into affecting 55 other methods, as well as methods outside the BHoM_Engine. But in general, I would not be against it. Wrapping the split segments into single segment polycurves does not feel that pretty, fully agree on that.

Will check what effect a change of return type has on the engine to start with.

Reply:
Seems fine to change return type, so have actioned it and updated the UTs for SplitAtPoint to comply. Seemed as most methods were calling the interface method anyway, hence no need to change the code elsewhere, and fully agree that it is better to just return single arcs/lines rather than them wrapped in a polycurve for no real reason :)